### DISCUSSION TITLE
Running multiple websites on a single server

### DISCUSSION THREAD

**Discussion by Juddling** [BretFisher/ama#151] [Q&A]
Thanks for your efforts on your course Bret, I've really enjoyed all your videos so far.

Currently the web agency I work for might have a production server which is set up with ~10 projects all using the globally installed nginx, PHP, MySQL etc.

I want each project to have its own isolated setup which we can use for development and production but I'm concerned with the overheads of having a 10 MySQL containers running on a server, 10 nginx containers, etc.

Are my concerns valid? Or should I perhaps be looking to set up my deployment in a different way?

**Comment by BretFisher** [comment]
This is the same concern you should have for any proper sizing of apps. This is a "normal" setup in many environments, so keep on doing what you're doing... but yes of course you need to have a handle on your resources.

You'll want to use reservations and constraints on your mysql containers to ensure they don't fight over resources. These work in docker run or in swarm services.

You'll want to understand your memory/cpu/disk/network i/o requirements (or at least start monitoring and measuring them) so you can ensure the server has the resources you need to run all those workloads.

In general, the more work you put on a single OS/server, the more you'll need to manage system settings on that server to get optimum performance. That's not docker specific, it's just that in the pre-docker cloud VM world you might be used to "a lot of small VM's with one app each", but now with "lots of apps on a single VM" you'll need to spend your time at the OS level, where before you might spend your time at the VM-management level.

Hope that makes sense. There is no silver bullet. You'll need to try running things and make adjustments many times to get to a system that both performs the way you want and also doesn't break the bank (obviously giving a single service 2-3x the hardware needed is one way to solve the problem, but you may not have the $$ to do that)

**Comment by BretFisher** [comment]
Keeping this open for visibility so others can learn too!
